miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef ADMFEM_HPP 
+#define ADMFEM_HPP
+
+#include "mfem.hpp"
+#include "fdual.hpp"
+#include "tadvector.hpp"
+#include "taddensemat.hpp"
+
+#ifdef MFEM_USE_CODIPACK
+#include <codi.hpp>
+
+namespace mfem {
+
+namespace ad {
+#ifdef MFEM_USE_ADFORWARD
+typedef codi::RealForward ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#else
+typedef codi::RealReverse ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#endif
+}
+
+
+template<int vector_size=1, int state_size=1, int param_size=0>
```

Comment:
As users would want to directly interface with this class, it needs doxygen.

Reply:
Including the template parameters.

Reply:
I suggest that the class documentation should include a link from the term automatic differentiation to something like the Wikipedia article on the subject: https://en.wikipedia.org/wiki/Automatic_differentiation

Reply:
documented